### PR TITLE
Update FeeSharingCollector to withhold part of fees to the protocol

### PR DESCRIPTION
**Pull request by tjcloa**
FeeSharingCollector should withhold a portion of fees for the protocol in certain tokens

### CODE REVIEW

File: deployment/deploy/1020-deploy-FeeSharingCollector.js [hunk 1]
```diff
@@ -18,6 +20,61 @@ const func = async function (hre) {
...
+    const wrbtcToken = (await get("WRBTC")).address;
+    const zusdDeployment = (await getOrNull("ZUSDToken")) || (await getOrNull("ZUSD"));
+    if (!zusdDeployment) {
+        throw new Error("Unable to resolve ZUSD token deployment (expected ZUSDToken or ZUSD)");
+    }
+    const zusdToken = zusdDeployment.address;
+    const rbtcDummyAddress = await feeSharingCollector.RBTC_DUMMY_ADDRESS_FOR_CHECKPOINT();
+
+    const withholdTokens = [
+        { name: "RBTC_DUMMY_ADDRESS_FOR_CHECKPOINT", address: rbtcDummyAddress },
```
Comment (cwsnt) [inline deployment/deploy/1020-deploy-FeeSharingCollector.js:45]: Adding WRBTC to withhold list has zero effects, because wrbtc will always be converted to `RBTC_DUMMY_ADDRESS_FOR_CHECKPOINT` [here](https://github.com/DistributedCollective/Sovryn-smart-contracts/blob/6913b7164351d0088a6d1c93df66d807804be67f/contracts/governance/FeeSharingCollector/FeeSharingCollector.sol#L229)

File: contracts/governance/FeeSharingCollector/FeeSharingCollector.sol [hunk 2]
```diff
@@ -207,7 +150,7 @@ contract FeeSharingCollector is
                 "FeeSharingCollector::withdrawFees: wrbtc token amount exceeds 96 bits"
             );
 
-            _addCheckpoint(RBTC_DUMMY_ADDRESS_FOR_CHECKPOINT, amount96);
+            _addCheckpointOrWithholdProtocolFee(RBTC_DUMMY_ADDRESS_FOR_CHECKPOINT, amount96);
```
Comment (cwsnt) [inline contracts/governance/FeeSharingCollector/FeeSharingCollector.sol]: In [here](https://github.com/DistributedCollective/Sovryn-smart-contracts/blob/6913b7164351d0088a6d1c93df66d807804be67f/contracts/governance/FeeSharingCollector/FeeSharingCollector.sol#L160)
When RBTC_DUMMY is in the withhold list, no staker checkpoint is created, yet both events still fire as if a normal fee distribution occurred.
Not sure if this will cause any issues with any off-chain indexer we have?
I'd suggest to suppress these events when withheld, or maybe add a flag to the event (bool withheld)

File: tests/FeeSharingCollectorTokenWithholdingTest.js [hunk 3]
```diff
@@ -0,0 +1,742 @@
...
+                secondAmount.toString(),
+                "withheld fees should include unprocessed amount"
+            );
+
+            totalCheckpoints = await feeSharingCollector.totalTokenCheckpoints(SOVToken.address);
+            assert.equal(totalCheckpoints.toString(), "1", "Should not create new checkpoint");
+        });
+    });
+
+    describe("Withdraw Protocol Withheld Fees", () => {
```
Comment (cwsnt) [inline tests/FeeSharingCollectorTokenWithholdingTest.js:623]: Seems like we are missing `withdrawFees()` and `withdrawFeesAMM()` test case scenario with the withheld token.

File: contracts/governance/FeeSharingCollector/FeeSharingCollector.sol [hunk 4]
```diff
@@ -1095,6 +1051,169 @@ contract FeeSharingCollector is
...
     }
 
+    /**
+     * @notice Add a token to the protocol withhold list.
+     * @dev Only callable by owner.
+     * @dev Fees from tokens in this list are withheld 100% by protocol (no staker distribution).
+     * @dev If there are unprocessed fees accumulated before adding to list, they are moved to protocolWithheldFees.
+     * @param token The token address to add.
+     */
+    function addProtocolWithholdToken(address token) external onlyOwner {
```
Comment (cwsnt) [inline contracts/governance/FeeSharingCollector/FeeSharingCollector.sol:1058]: unprocessedAmount is correctly moved to protocolWithheldFees. But lastFeeWithdrawalTime[token] is left as-is. While the tkoen is withheld, no one updates lastFeeWithdrawalTime. If the token is later removed from the withhold list and fees start flowing again, the interval may still be "running" from before the token was ever added to the list. The first staker checkpoint could be delayed by a stale interval or fire immediately depending on when things happened.
Do youy  think we should consider resetting lastFeeWithdrawalTime[token] = 0 when adding to the list so the staker flow restarts cleanly if the token is ever removed?

File: contracts/governance/IFeeSharingCollector.sol [hunk 5]
```diff
@@ -5,9 +5,99 @@ pragma solidity ^0.5.17;
...
+    /// @notice An event emitted when protocol withheld fees are withdrawn.
+    event ProtocolWithheldFeesWithdrawn(
+        address indexed sender,
+        address indexed receiver,
+        address indexed token,
+        uint256 amount
+    );
+
+    /// @notice An event emitted when revenue is accumulated.
+    event ProtocolRevenueAccumulated(address indexed token, uint256 amount);
```
Comment (cwsnt) [inline contracts/governance/IFeeSharingCollector.sol:80]: should we add the sender param here?
```event ProtocolRevenueAccumulated(address indexed sender, address indexed token, uint256 amount);```

File: contracts/governance/FeeSharingCollector/FeeSharingCollector.sol (hunk 2 above)
Comment (tjcloa) [reply contracts/governance/FeeSharingCollector/FeeSharingCollector.sol]: `emit FeeWithdrawnInRBTC(msg.sender, wrbtcAmountWithdrawn);` is fired in any event indicating that the fees were withdrawn from the protocol no matter how it is distributed further.  
if a checkpoint is created, the [respecitve event](https://github.com/DistributedCollective/Sovryn-smart-contracts/blob/6913b7164351d0088a6d1c93df66d807804be67f/contracts/governance/FeeSharingCollector/FeeSharingCollector.sol#L983) emitted. otherwise it is withheld to the protocol and the [ProtocolRevenueAccumulated](https://github.com/DistributedCollective/Sovryn-smart-contracts/blob/6913b7164351d0088a6d1c93df66d807804be67f/contracts/governance/FeeSharingCollector/FeeSharingCollector.sol#L264) event is emitted.  
if the subgraph relies only on the `FeeWithdrawnInRBTC` considering it a distribution event, it should be changed because it mixes up withdraw with distribution.

File: contracts/governance/FeeSharingCollector/FeeSharingCollector.sol (hunk 4 above)
Comment (tjcloa) [reply contracts/governance/FeeSharingCollector/FeeSharingCollector.sol:1058]: good catch. updated, added tests.

File: contracts/governance/FeeSharingCollector/FeeSharingCollector.sol [hunk 6]
```diff
@@ -1095,6 +1051,169 @@ contract FeeSharingCollector is
...
+     */
+    function removeProtocolWithholdToken(address token) external onlyOwner {
+        require(
+            protocolWithheldTokensList.contains(token),
+            "FeeSharingCollector::removeProtocolWithholdToken: token not in list"
+        );
+
+        protocolWithheldTokensList.remove(token);
+        emit TokenRemovedFromProtocolWithholdList(msg.sender, token);
+    }
```
[the remark this answers is not in the document]
Comment (tjcloa) [reply contracts/governance/FeeSharingCollector/FeeSharingCollector.sol:1095]: added this info to the funcion natspec @notice

File: contracts/governance/IFeeSharingCollector.sol (hunk 5 above)
Comment (tjcloa) [reply contracts/governance/IFeeSharingCollector.sol:80]: why do you think it matters who executed the function?

File: deployment/deploy/1020-deploy-FeeSharingCollector.js (hunk 1 above)
Comment (tjcloa) [reply deployment/deploy/1020-deploy-FeeSharingCollector.js:45]: removed WRBTC

File: deployment/deploy/1020-deploy-FeeSharingCollector.js [hunk 7]
```diff
@@ -18,6 +20,61 @@ const func = async function (hre) {
...
+        { name: "ZUSD", address: zusdToken },
+        { name: "WRBTC", address: wrbtcToken },
```
[the remark this answers is not in the document]
Comment (tjcloa) [reply deployment/deploy/1020-deploy-FeeSharingCollector.js]: removed WRBTC

File: tests/FeeSharingCollectorTokenWithholdingTest.js [hunk 8]
```diff
@@ -0,0 +1,742 @@
...
+            );
+            assert.equal(
+                totalCheckpoints.toString(),
+                "0",
+                "Should not create checkpoint without transfers"
+            );
+
+            // The actual zero-value protection is in _addCheckpoint:
+            // if (amount > 0) { _writeTokenCheckpoint(_token, amount); }
+            // This prevents creating a checkpoint when accumulated amount is 0
```
[the remark this answers is not in the document]
Comment (tjcloa) [reply tests/FeeSharingCollectorTokenWithholdingTest.js]: removed comment

File: tests/FeeSharingCollectorTokenWithholdingTest.js [hunk 9]
```diff
@@ -0,0 +1,742 @@
...
+            );
+            assert.equal(
+                totalCheckpoints.toString(),
+                "0",
+                "Should not create checkpoint with zero accumulated amount"
+            );
+
+            // Note: transferTokens and transferRBTC both require amount > 0 in their input validation,
+            // so they won't call _addCheckpoint with 0. The protection in _addCheckpoint handles
+            // edge cases where unprocessedAmount could theoretically accumulate to 0.
```
[the remark this answers is not in the document]
Comment (tjcloa) [reply tests/FeeSharingCollectorTokenWithholdingTest.js:752]: updated

File: tests/FeeSharingCollectorTokenWithholdingTest.js (hunk 3 above)
Comment (tjcloa) [reply tests/FeeSharingCollectorTokenWithholdingTest.js:623]: added above

File: contracts/governance/IFeeSharingCollector.sol (hunk 5 above)
Comment (cwsnt) [reply contracts/governance/IFeeSharingCollector.sol:80]: Was thinking it's for consistency, since other event is also storing this `sender` as well,
`FeeWithdrawnInRBTC` for example

But i will leave it to u as optional

**Review by cwsnt — APPROVED** [review]
